core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java

```diff
@@ -79,20 +134,137 @@ public void stop() throws MuleException {
     if (processorChain != null) {
       processorChain.stop();
     }
+    stopIfNeeded(processingStrategy);
   }
 
   @Override
   public CoreEvent process(CoreEvent event) throws MuleException {
-    return processorChain.process(event);
+    return processToApply(event, chainWithPs);
   }
 
   @Override
   public Publisher<CoreEvent> apply(Publisher<CoreEvent> publisher) {
-    return from(publisher)
+    return Mono.from(publisher)
         .doOnNext(notificationHelper.notification(PROCESS_START))
-        .transform(processorChain)
+        .transform(chainWithPs)
         .doOnSuccess(notificationHelper.notification(PROCESS_END))
         .doOnError(MessagingException.class, notificationHelper.errorNotification(PROCESS_END));
   }
 
+  /**
+   * Will execute each of the 2 segments of a policy chain (the before and after segments relative to the {@code execute-next}
+   * component) in its own scheduler, based on the 'heaviest' {@link ProcessingType} of its processors.
+   */
+  private static final class SourcePolicyProcessingStrategy extends AbstractProcessingStrategy
+      implements ProcessingStrategy, Startable, Stoppable {
+
+    private SchedulerService schedulerService;
+    private SchedulerConfig schedulerBaseConfig;
+    private String schedulersNamePrefix;
+    private List<Processor> beforeNext;
+    private List<Processor> afterNext;
+
+    private Scheduler beforeScheduler;
+    private Scheduler afterScheduler;
+    private Scheduler cpuLiteBeforeScheduler;
+    private Scheduler cpuLiteAfterScheduler;
+
+    public SourcePolicyProcessingStrategy(SchedulerService schedulerService, SchedulerConfig schedulerBaseConfig,
+                                          String schedulersNamePrefix, List<Processor> beforeNext, List<Processor> afterNext) {
+      this.schedulerService = schedulerService;
+      this.schedulerBaseConfig = schedulerBaseConfig;
+      this.schedulersNamePrefix = schedulersNamePrefix;
+      this.beforeNext = beforeNext;
+      this.afterNext = afterNext;
+    }
+
+    @Override
+    public void start() throws MuleException {
+      beforeScheduler = schedulerBasedOnProcessingTypes(beforeNext, schedulersNamePrefix + "-before");
```

Comment:
If before does IO and so does after, will this mean two different schedulers? Is there any way to avoid/simplify this?

Reply:
there is no cost ins having 2 schedulers, the actual thread pool is the same.